models/clm/tutorial/README.rst

```diff
@@ -0,0 +1,1567 @@
+| Settings                    |                                                             |
++=============================+=============================================================+
+| ``DATA_ASSIMILATION_LND``   | If ``.TRUE.`` data assimilation is enabled.  If ``.FALSE.`` |
+|                             | no data assimilation is performed, and simulation is        |
+|                             | performed as a normal CLM run.                              |
++-----------------------------+-------------------------------------------------------------+
+| ``DATA_ASSIMILATION_SCRIPT``| Location of script ``assimilate.csh`` that performs the     |
+|                             | assimilation.  Controls execution of DART scripts and passes|
+|                             | files between DART and CLM code.                            |
++-----------------------------+-------------------------------------------------------------+
+| ``STOP_OPTION``             | Unit of time that controls duration of each assimilation    |
+|                             | cycle (see ``STOP_N``)                                      |
++-----------------------------+-------------------------------------------------------------+
+| ``STOP_N``                  | Duration of each assimilation cycle in units of time        |
+|                             | defined by ``STOP_OPTION``                                  |
++-----------------------------+-------------------------------------------------------------+
+| ``DATA_ASSIMILATION_CYCLES``| The number of assimilation cycles performed for a single    |
+|                             | job submission                                              |
++-----------------------------+-------------------------------------------------------------+
+| ``RESUBMIT``                | The number of times each job submission is repeated as      |
+|                             | defined by ``STOP_OPTION``, ``STOP_N``, and                 |
+|                             | ``DATA_ASSIMILATION_CYCLES``                                |
++-----------------------------+-------------------------------------------------------------+
+| ``CONTINUE_RUN``            | If ``.FALSE.`` this is the startup of the assimilation run  |
+|                             | beginning from the ``refcase``.  If ``.TRUE.`` this is a    |
+|                             | continuation of a previous assimilation run.                |
++-----------------------------+-------------------------------------------------------------+ 
+
+ 
+Before submission review your ``input.nml`` within your case folder to confirm
+the settings reflect those from previous steps of the tutorial.  Next submit
+the assimilation run:
+
+::
+
+ > cd <caseroot>
+ > ./case.submit
+
+Check the status of the job on Cheyenne using PBS commands to determine if job
+is queued (Q), running (R) or completed.
+
```

Comment:
I'm at this stage at the moment. My filter is failing at the `fill_inflation_restart` step. I'm not sure if I did something wrong or we have to fix something. Here is the error: 
```  
fill_inflation_restart:  - Reading File : clm_restart.nc
  fill_inflation_restart:  - Reading File : clm_history.nc
 ERROR FROM:
  source : netcdf_utilities_mod.f90
  routine: read_variables: nf90_get_var
  message: TLAI: NetCDF: Start+count exceeds dimension bound
  message: ...
```

Reply:
Thanks for working through this @mgharamti -- can you provide your ``input.nml`` file within your ``<caseroot>`` directory that led to the ``fill_inflation_restart`` error?

Reply:
Brett, here is my case directory: `/glade/work/gharamti/cases/cesm_dart/clm5_dart_tut` The input.nml file should be in there.

Reply:
Looking at this now-- problem is definitely on my end.

Reply:
So that took longer to figure out than I expected -- sorry about that.  The problem is that the DART state is missing the clm variable for soil temperature ``TSOI``.   Adding this variable into the ``&model_nml`` should fix the problem, and should look like:

```
   clm_variables  = 'leafc',       'QTY_LEAF_CARBON',            '0.0', 'NA', 'restart' , 'UPDATE',
                    'frac_sno',    'QTY_SNOWCOVER_FRAC',         '0.0', '1.', 'restart' , 'NO_COPY_BACK',
                    'SNOW_DEPTH',  'QTY_SNOW_THICKNESS',         '0.0', 'NA', 'restart' , 'NO_COPY_BACK',
                    'H2OSOI_LIQ',  'QTY_SOIL_LIQUID_WATER',      '0.0', 'NA', 'restart' , 'UPDATE',
                    'H2OSOI_ICE',  'QTY_SOIL_ICE',               '0.0', 'NA', 'restart' , 'UPDATE',
                    'T_SOISNO',    'QTY_TEMPERATURE',            '0.0', 'NA', 'restart' , 'UPDATE',
                    'livestemc',   'QTY_LIVE_STEM_CARBON',       '0.0', 'NA', 'restart' , 'UPDATE',
                    'deadstemc',   'QTY_DEAD_STEM_CARBON',       '0.0', 'NA', 'restart' , 'UPDATE',
                    'TSOI',        'QTY_SOIL_TEMPERATURE',       'NA' , 'NA', 'history' , 'NO_COPY_BACK',
                    'TLAI',        'QTY_LEAF_AREA_INDEX',        '0.0', 'NA', 'vector'  , 'NO_COPY_BACK',
   /
```

This was my fault for not double-checking the run configuration before issuing the PR, sorry about that.  The error code for this was a bit misleading as well -- that is because not only is the clm variable ``TSOI`` used as a forward operator for the assimilation of soil temperature, but also ``TSOI`` is within the ``history`` file domain for clm.  The ``history`` file is a unique domain to DART.  Therefore the code was looking for the 3 domains including ``restart``, ``history`` and ``history_vector``, but the clm_variables list only specified 2 out of 3 of those domains, which triggered the error.  Should work now.    Will add a commit to permanently fix this.

Reply:
@braczka Thanks for checking this. Here is what's happening now. `fill_inflation_restart` worked fine but `filter` failed with this error

``` 
Before preassim state space output TIME: 2022/03/08 16:47:02
 ERROR FROM:
  source : netcdf_utilities_mod.f90
  routine: write_variables:
  message: nf90_inq_varid "leafc": NetCDF: Variable not found
  message: ...
 
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI COMMUNICATOR 3 DUP FROM 0
with errorcode 99.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
MPT ERROR: could not run executable.
	(HPE MPT 2.21  11/28/19 04:22:45)
Killed
```

The variable `leafc` seems to be present in my prior ensemble members `clm5_dart_tut.clm2_000?.r.2011-01-02-00000.nc`! I'm not sure what's triggering this. Here is my run directory if you'd like to have a look: `/glade/scratch/gharamti/cesm_dart/clm5_dart_tut/run`

Reply:
I compared both our tutorial cases on Cheyenne and they look identical in terms of the the setup scripts of ``CLM5_setup_assimilation``, ``DART_params.csh`` and ``input.nml``.    Not obvious what is going on so I diffused both of our ``da.log`` files:   ``/glade/scratch/bmraczka/cesm2.2.0/clm5_SWE0_tutorialfix/run/da.log.3199371.chadmin1.ib0.cheyenne.ucar.edu.220308-141434`` against ``/glade/scratch/gharamti/cesm_dart/clm5_dart_tut/run/da.log.3201989.chadmin1.ib0.cheyenne.ucar.edu.220308-164401``  and there seems to be an important difference in how the mpi process are allocated during ``filter``.  For example, I show below excerpted differences in our da.log files:

```
 --------------------------------------
 Starting ... at YYYY MM DD HH MM SS = 
                 2022  3  8 14 17 22
 Program Filter
 --------------------------------------
 
  set_nml_output Echo NML values to log file only
 PE 0: initialize_mpi_utilities:  Running with          360  MPI processes.
 PE 0: static_init_model  assimilation period is            2  days         3600
   seconds
```
compared to:

```
 --------------------------------------
 Starting ... at YYYY MM DD HH MM SS = 
                 2022  3  8 16 46 51
 Program Filter
 --------------------------------------
 
  set_nml_output Echo NML values to log file only
 PE 0: initialize_mpi_utilities:  Running with            1  MPI processes.
 PE 0: static_init_model  assimilation period is            2  days         3600  seconds
```

Also my da.log provides the following chunk of lines, where yours does not:

```
 PE 0: assim_tools_init: Replicating a copy of the ensemble mean on every task
 PE 0: assim_tools_init: ... uses more memory per task but may run faster if doi
 ng vertical
 PE 0: assim_tools_init: ... coordinate conversion; controlled by namelist item 
 "distribute_mean"
 PE 0: assim_tools_init: Not doing vertical localization, no vertical coordinate
  conversion required
```

 Finally yours gives an mpt error during ``filter`` prior to the ``leafc`` error:
```
 PE 0: create_and_open_state_output  Creating output file preassim_member_0001_d01.nc
MPT: Launch error on r14i1n12.ib0.cheyenne.ucar.edu
```

Not sure what accounts for these differences, it looks like our run log files submit the jobs identically:
```
run command is mpiexec_mpt -p "%g:"  -np 360  omplace -tm open64  /glade/scratch/bmraczka/cesm2.2.0/clm5_SWE0_tutorialfix/bld/cesm.exe   >> cesm.log.$LID 2>&1
   Running /glade/work/bmraczka/cases/cesm2.2.0/clm5_SWE0_tutorialfix/assimilate.csh

```


I don't set up my DART environment in any particular way.  The modules loaded on Cheyenne are:

```
Currently Loaded Modules:
  1) ncarenv/1.3   2) intel/19.0.5   3) ncarcompilers/0.5.0   4) mpt/2.22   5) netcdf/4.7.4   6) ncl/6.6.2   7) nco/4.9.5   8) diffuse/0.4.8
```

  I will continue to look at this..... maybe @hkershaw-brown can take a look too?

Reply:
For leafc not being found: Brett's log file shows the preassim files being created. 

Moha's log does not, so is Moha's case trying to write to an existing file that does not have leafc? Maybe a file from a previous run?

For the extra log message "Replicating a copy", Brett is getting that because he is running on 360 processes and Moha is running on 1. For 1 process there is no need to warn about distributing the state across processes. So the extra log message is expected - but the running on 1 process means something is not quite right. 

I think Moha has a funky mix of openmpi and mpt in his environment.  The error message at the end of Moha's log file is 

```diff
MPI_ABORT was invoked on rank 0 in communicator MPI COMMUNICATOR 3 DUP FROM 0
with errorcode 99.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
-exactly when Open MPI kills them.
--------------------------------------------------------------------------
-MPT ERROR: could not run executable.
	(HPE MPT 2.21  11/28/19 04:22:45)
Killed
```

I think this is weird.  I would have a look at the mpi first, then the netcdf problem. 
For example,  if you have 360 serial versions of filter trying to write to the same netcdf file at the same time might cause unexpected errors. I'm not saying this is the cause of the netcdf problem here, but you can get funky effects e.g.
https://github.com/NCAR/DART/issues/308

Brett, I'll give running the tutorial a go with the default modules. Maybe worth Moha doing a `module purge`, `module load blah blah blah` before running the tutorial. I think the CESM case is set up to use MPT (correct me if I wrong).

Reply:
Thanks Helen -- will be helpful to see what you get when you run the tutorial.  The only significant difference that Moha and I could identify between our modules was I had `intel/19.0.5 ` loaded whereas Moha had `intel/19.1.1`.   Doesn't seem the likely difference.   Also our CESM and DART versions were identical......

Reply:
Ran the tutorial, runs the 5 assimilation cycles to completion.  Runs filter with 360 processes
/glade/scratch/hkershaw/cesm_dart/clm5_f09_assim_e5/run/

Reply:
I removed my cesm_dart directory and recompiled both dart and the model. The experiment completed the 5 cycles successfully. Here is my run: `/glade/work/gharamti/cases/cesm_dart/clm5_dart_tut` 
I think when I switched from `openmpi` to `mpt` I only re-complied one executable (cannot recall whether it was DART or CESM). So, it's good now. 
Now, I want to re-iterate my previous my comment. I think it's important to discuss the modules before asking the users to compile. I usually use `openmpi` (for my other application). I went through the tutorial without even realizing the need for `mpt`. I got an error during the run and then I made a swap in the modules. This needs to be clarified in STEP 1 or 2.

Reply:
@hkershaw-brown thanks for your insights again!

Reply:
Thanks @hkershaw-brown and very much appreciate the feedback @mgharamti.   Will make a point to document/discuss the module environment before the user compiles.